### ISSUE TITLE
GPU acceleration problems (again)

### ISSUE THREAD

**Issue by Hellsy22** [alexjc/neural-doodle#57]
I'm totaly new with CUDA and GPU computations.
Trying to use GeForce GTX 950 with 2Gb
After ~3hrs it failed:
```
# python3 doodle.py --style samples/Renoir.jpg --output samples/Reactor.png --device=gpu0 --iterations=80

  - Using device `gpu0` for processing the images.
  - Loading content semantic map from samples/Reactor_sem.png.
  - Loading style image data from samples/Renoir.jpg.
  - Loading style semantic map from samples/Renoir_sem.png.
...
Phase #2: resolution 878x770  scale 1.0
  - Style layer sem3_1: 9828 patches in 22,909kb.
  - Style layer sem4_1: 2356 patches in 10,920kb.

Error allocating 1557642240 bytes of device memory (out of memory). Driver report 1307291648 bytes free and 2146762752 bytes total 
ERROR: The optimization diverged and NaN numbers were encountered.
  - Try using a different device or change the parameters.
```

Before crush:
```+------------------------------------------------------+                       
| NVIDIA-SMI 361.28     Driver Version: 361.28         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 950     On   | 0000:01:00.0      On |                  N/A |
|  0%   52C    P8    11W / 110W |    283MiB /  2047MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      5977    G   /usr/bin/X                                      51MiB |
|    0      6854    G   ...s-passed-by-fd --v8-snapshot-passed-by-fd    25MiB |
|    0      8080    C   python3                                        191MiB |
+-----------------------------------------------------------------------------+
```

1. It shouldn't take so long, right? What should I check & fix?
2. I installed libcudnn and set env var LD_LIBRARY_PATH, but how can I check it working at all?

**Comment by alexjc** [comment]
I think Memory is a separate issue, but it seems way too slow for GPU. Make sure $CUDA_PATH is setup before you install Theano. See comments here: https://github.com/alexjc/neural-doodle/issues/51

`nvidia-smi dmon` should show your GPU usage, it should show spikes at 40-50 percent usage.

**Comment by Hellsy22** [comment]
I never used CUDA or Theano before, so I wrote all the steps to check it:

CUDA installed at /opt/cuda, so .bashrc got:
```bash
export PATH=/opt/cuda/bin:$PATH
export CUDA_ROOT=/opt/cuda/bin
export CUDA_PATH=/opt/cuda/bin
export LD_LIBRARY_PATH=/opt/cuda/lib:/opt/cuda/lib64:$LD_LIBRARY_PATH
```
for example, ```nvcc``` located at ```/opt/cuda/bin/```, ```libcudnn``` at ```/opt/cuda/lib64```

Then I reinstalled systemwide numpy and scipy. Deleted neural-doodle and re-cloned it from GIT, reinstalled packages with
```bash
python3 -m venv pyvenv --system-site-packages
source pyvenv/bin/activate
python3 -m pip install --ignore-installed -r requirements.txt
```

I can't delete ```pyvenv``` folder, because without ```source pyvenv/bin/activate```  neural-doodle fail with ```ImportError: No module named 'skimage'```

Rebooted to ensure all modules loaded.

Looks like it solves the speed problem. So, if previous part is correct let's see a memory problem:

[Full output at pastebin](http://pastebin.com/CHGFcdhR)
How can I check of using ```libcudnn```?
How can I create Theano config and should I? If yes, can you provide filename/location and example, please?

**Comment by alexjc** [comment]
Great news! I'll add these steps to the Troubleshooting section tomorrow.

The memory problem won't go away without some new code to split up the rendering and use less memory. You could also do the first few resolutions on GPU then switch to CPU.

**Comment by Hellsy22** [comment]
I'm not sure about steps. May be some of them unnecessary or even wrong. May be it's only because of Gentoo and pure installations of cuda/nvidia-drivers will be different. It worked for me but I even don't know is ```libcudnn``` in use or not.

About memory - what should be smaller - input samples, output or both?

**Comment by alexjc** [comment]
You can set `THEANO_FLAGS=device=gpu0,floatX=float32` and it should display a bit more information about the driver.

Output has bigger impact on memory usage, but input determines the number of patches which also need to fit in memory. DeepForger uses some tiling code for rendering, I'll see about bringing that over!

**Comment by Hellsy22** [comment]
Downscaled output ```_sem``` to 0.7,  used flags like this:
```THEANO_FLAGS=device=gpu0,floatX=float32 python3 doodle.py --style samples/Renoir.jpg --output samples/Reactor.png --device=gpu0 --iterations=80```

Got this info:
Using gpu device 0: GeForce GTX 950 (CNMeM is disabled, **CuDNN 4007**)
Looks like 600x600 is a cap for 2Gb

**Comment by alexjc** [comment]
Yep, that looks correct. CUDNN is already saving you memory so 600x600 is not bad!

I will push more memory saving changes soon once the current experiment improving quality is done.
